### ISSUE TITLE
'Missing return statement' error occurring in the match statement in Python 3.10.

### ISSUE THREAD

**Issue by Supertonian** [python/mypy#14704]
**Bug Report**

(A clear and concise description of what the bug is.)

**To Reproduce**

```python
# Ideally, a small sample program that demonstrates the problem.
# Or even better, a reproducible playground link https://mypy-play.net/ (use the "Gist" button)

def to_bytes(value: Any) -> bytes:
    match value:
        case s if isinstance(s, str):
            return s.encode("utf-8")
        case b if isinstance(b, bytes):
            return b
        case _:
            raise ValueError('value cannot be expressed in bytes')
```

**Expected Behavior**

It should not raise error 'Missing return statement'

**Actual Behavior**

Raises error 'Missing return statement'.

Even changing raise inside case_ to return b"" still produces 'Missing return' error.

**Your Environment**

- Mypy version used: 1.0.0
- Mypy command-line flags: mypy <filename>
- Mypy configuration options from `mypy.ini` (and other config files):
```
[tool.mypy]
disallow_any_generics = true
disallow_untyped_defs = true
exclude = ['pb2\.py$', 'pb2_grpc\.py$']
files = ["src", "tests"]
follow_imports = "silent"
ignore_missing_imports = true
python_version = "3.10"
```
- Python version used: 3.10.9

**Comment by henryiii** [comment]
It's unhappy with the guard:

```python
def f(x: object) -> bool:
    match x:
        case {"a": a} if a:
            return True
        case _:
            return False
```

```console
$ mypy --strict tmp.py
tmp.py:1: error: Missing return statement  [return]
Found 1 error in 1 file (checked 1 source file)
```

Mypy works if you remove the `if a`, but claims there is a missing return statement when that `if a` is present. Guards just allow fall through, so the final case does always trigger if the one above doesn't, there's no missing return statement.

**Comment by Supertonian** [comment]
Is there a way to perform type checking using the match statement, without resorting to using the isinstance function inside the case statement? @henryiii

**Comment by henryiii** [comment]
Not sure what that has to do with the missing return statement bug, but yes - `match MyClass()` (notice the parentheses) replaces `isinstance`. That was the motivating reason to introduce pattern matching, as it could replace most `isinstance` uses. (I guess knowing this does help avoid this bug if you are using `isinstance` in a guard, which shouldn't be necessary?)

**Comment by ilmari** [comment]
This looks like a duplicate of #12534
